### PR TITLE
secret: added package unit tests

### PR DESCRIPTION
**Pull request by eldar101**
Unittests for the secret package.

### CODE REVIEW

File: pkg/secret/secret_test.go [hunk 1]
```diff
@@ -1 +1,353 @@
...
+	testCases := []struct {
+		secretExistsAlready bool
+		name                 string
+		namespace            string
+	}{
+		{
+			secretExistsAlready: true,
+			name:                 "test",
+			namespace:            "test",
+			secretType:           "test",
```
Comment (achuzhoy) [inline pkg/secret/secret_test.go:153]: redunant.

File: pkg/secret/secret_test.go [hunk 2]
```diff
@@ -1 +1,353 @@
...
+		{
+			expectedError:       true,
+			addToRuntimeObjects: true,
+			expectedErrorText:  "secret namespace cannot be empty",
+			secretName:         "test",
+			secretNamespace:    "",
+			secretType:         "test", 
+
+		},
+		{
```
Comment (achuzhoy) [inline pkg/secret/secret_test.go:54]: Redundant?

File: pkg/secret/secret_test.go (hunk 1 above)
Comment (achuzhoy) [reply pkg/secret/secret_test.go]: just secretType

**Comment by kononovn** [comment]
@eldar101 Please re-base. There is new function in secret package

File: go.mod [hunk 3]
```diff
@@ -27,6 +27,7 @@ require (
 	github.com/openshift/local-storage-operator v0.0.0-20240422172451-2a80d7f6681d // release-4.16
 	github.com/openshift/machine-config-operator v0.0.1-0.20230811181556-63d7be1ef18b
 	github.com/openshift/ptp-operator v0.0.0-20240404165119-29a3d7b3d60b
+	github.com/openshift-kni/eco-goinfra latest
```
Comment (kononovn) [inline go.mod:30]: Please remove this line from go.mod

File: pkg/secret/secret_test.go [hunk 4]
```diff
@@ -1 +1,347 @@
...
+		result := testBuilder.Exists()
+		if testCase.secretExistsAlready {
+			assert.True(t, result)
+		} else {
+			assert.False(t, result)
+		}
+	}
+}
+
+
```
Comment (kononovn) [inline pkg/secret/secret_test.go:330]: Please add test cases for following functions: NewBuilder, WithOptions, WithAnnotations, WithData, Update

**Review by kononovn — CHANGES_REQUESTED** [review]
Please check few comments above.

File: pkg/secret/secret_test.go [hunk 5]
```diff
@@ -1 +1,502 @@
...
+			secretName:         "test",
+			secretNamespace:    "",
+		},
+		{
+			expectedError:       true,
+			addToRuntimeObjects: true,
+			expectedErrorText:  "secret type cannot be empty",
+			secretName:         "test",
+			secretNamespace:    "test",
+			secretType:         "",
```
Comment (kononovn) [inline pkg/secret/secret_test.go:56]: I think we need to remove it.

File: pkg/secret/secret_test.go [hunk 6]
```diff
@@ -1 +1,502 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+func TestSecretPull(t *testing.T) {
+	testCases := []struct {
+		secretName         string
+		secretNamespace    string
+		expectedError       bool
+		addToRuntimeObjects bool
+		expectedErrorText   string
```
Comment (kononovn) [inline pkg/secret/secret_test.go:19]: Please add client bool args. Refer to: https://github.com/openshift-kni/eco-goinfra/blob/5966cfaaf0378411c7307befc09a20b1e59e1df2/pkg/metallb/adresspool_test.go#L44
and Add 1 more test case when client is false(nil)

File: pkg/secret/secret_test.go [hunk 7]
```diff
@@ -1 +1,502 @@
...
+				assert.Equal(t, testCase.expectedErrorText, err.Error())
+			}
+		} else {
+			assert.Nil(t, err)
+			assert.NotNil(t, builderResult)
+		}
+	}
+}
+
+func TestSecretCreate(t *testing.T) {
```
Comment (kononovn) [inline pkg/secret/secret_test.go:94]: Please add 1 more test case when secret has invalid definition.
Refer to :https://github.com/openshift-kni/eco-goinfra/blob/5966cfaaf0378411c7307befc09a20b1e59e1df2/pkg/metallb/adresspool_test.go#L212

File: pkg/secret/secret_test.go [hunk 8]
```diff
@@ -1 +1,502 @@
...
+		err := testBuilder.Delete()
+		assert.Nil(t, err)
+
+		// Assert that the object actually does not exist
+		_, err = Pull(client, testCase.name, testCase.namespace, testCase.secretType)
+		assert.NotNil(t, err)
+	}
+}
+
+func TestSecretValidate(t *testing.T) {
```
Comment (kononovn) [inline pkg/secret/secret_test.go:181]: This is private function. Please move it to the end of the file

File: pkg/secret/secret_test.go [hunk 9]
```diff
@@ -1 +1,502 @@
...
+			assert.Equal(t, testCase.expectedError, err.Error())
+			assert.False(t, result)
+		} else {
+			assert.Nil(t, err)
+			assert.True(t, result)
+		}
+	}
+}
+
+func TestSecretGet(t *testing.T) {
```
Comment (kononovn) [inline pkg/secret/secret_test.go:241]: Secret doesn't have Get function. Please remove it

File: pkg/secret/secret_test.go [hunk 10]
```diff
@@ -1 +1,502 @@
...
+		result := testBuilder.Exists()
+		if testCase.secretExistsAlready {
+			assert.True(t, result)
+		} else {
+			assert.False(t, result)
+		}
+	}
+}
+
+func TestSecrettWithOptions(t *testing.T) {
```
Comment (kononovn) [inline pkg/secret/secret_test.go:330]: Please fix typo: TestSecre**tt**WithOptions> TestSecretWithOptions

File: pkg/secret/secret_test.go [hunk 11]
```diff
@@ -1 +1,502 @@
...
+			key:         "key",
+			value:       "value",
+			expectedErr: "",
+		},
+		{
+			key:         "",
+			value:       "",
+			expectedErr: "'data' cannot be empty",
+		},
+	}
```
Comment (kononovn) [inline pkg/secret/secret_test.go:361]: Empty function. Please write test function

File: pkg/secret/secret_test.go [hunk 12]
```diff
@@ -1 +1,502 @@
...
+}
+
+func TestSecretNewBuilder(t *testing.T) {
```
Comment (kononovn) [inline pkg/secret/secret_test.go:364]: Please paste this function after TestSecretPull

File: pkg/secret/secret_test.go [hunk 13]
```diff
@@ -1 +1,502 @@
...
+		},
+		{
+			name:          "test-name",
+			namespace:     "",
+			secretType:    "test-secrettype",
+			expectedError: "secret 'nsname' cannot be empty",
+		},
+                {
+                        name:          "test-name",   
+                        namespace:     "",
```
Comment (kononovn) [inline pkg/secret/secret_test.go:391]: namespace is empty instead of secretType. Please fix

File: pkg/secret/secret_test.go [hunk 14]
```diff
@@ -1 +1,502 @@
...
+		},
+		{
+			testAnnotations:   map[string]string{"openshift.io/internal-registry-auth-token.service-account": "default"},
+			expectedErrorText: "",
+		},
+		{
+			testAnnotations:   map[string]string{},
+			expectedErrorText: "'annotations' argument cannot be empty",
+		},
+	}
```
Comment (kononovn) [inline pkg/secret/secret_test.go:429]: Test body is missing

File: pkg/secret/secret_test.go [hunk 15]
```diff
@@ -1 +1,502 @@
...
+}
+
+func TestUpdate(t *testing.T) {
```
Comment (kononovn) [inline pkg/secret/secret_test.go:432]: Please re-name to TestSecretUpdate

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix comments above

File: pkg/secret/secret_test.go (hunk 6 above)
Comment (kononovn) [reply pkg/secret/secret_test.go:26]: You added client but it's unused in test case.

Comment (kononovn) [reply pkg/secret/secret_test.go:26]: No, you have to add test case when you pass nil client to the Pull function and catch relevant error.
Please refer
https://github.com/openshift-kni/eco-goinfra/blob/5966cfaaf0378411c7307befc09a20b1e59e1df2/pkg/metallb/adresspool_test.go#L88

**Review by kononovn — CHANGES_REQUESTED** [review]
1 comment to fix

**Comment by kononovn** [comment]
@eldar101 Please change commit message and PR headline to:

`secret: added package unit tests`

**Comment by kononovn** [comment]
@eldar101 Please fix linter issues. Please run > make lint in order to make sure that basic check pass

**Review by kononovn — CHANGES_REQUESTED** [review]
please fix ci issues.

**Comment by kononovn** [comment]
@eldar101 Check still failed. Please fix.

File: pkg/secret/secret_test.go [hunk 16]
```diff
@@ -1 +1,516 @@
...
+		}
+
+		testBuilder, _ := buildTestBuilderWithFakeObjects(runtimeObjects, testCase.name, testCase.namespace)
+
+		result := testBuilder.Exists()
+		if testCase.secretExistsAlready {
+			assert.True(t, result)
+		} else {
+			assert.False(t, result)
+		}
```
Comment (trewest) [inline pkg/secret/secret_test.go:259]:
```suggestion
                assert.Equal(t, testCase.secretExistsAlready, testBuilder.Exists())
```

File: pkg/secret/secret_test.go [hunk 17]
```diff
@@ -1 +1,516 @@
...
+
+		if testCase.expectedErr == "" {
+			testBuilder.WithData(data)
+
+			assert.Equal(t, []byte(testCase.value), testBuilder.Definition.Data[testCase.key])
+		} else {
+			testBuilder.WithData(map[string][]byte{})
+
+			assert.Equal(t, testCase.expectedErr, testBuilder.errorMsg)
+		}
```
Comment (trewest) [inline pkg/secret/secret_test.go:313]:
```suggestion
		data[testCase.key] = []byte(testCase.value)
		
		testBuilder.WithData(data)
		assert.Equal(t, testCase.expectedErr, testBuilder.errorMsg)

		if testCase.expectedErr == "" {
			assert.Equal(t, []byte(testCase.value), testBuilder.Definition.Data[testCase.key])
		}
```

File: pkg/secret/secret_test.go [hunk 18]
```diff
@@ -1 +1,516 @@
...
+		testBuilder, _ := buildTestBuilderWithFakeObjects(runtimeObjects, testCase.Name, testCase.Namespace)
+
+		// Assert the secret before the update
+		assert.NotNil(t, testBuilder.Definition)
+
+		assert.Equal(t, testCase.Name, testBuilder.Definition.Name)
+		assert.Equal(t, testCase.Namespace, testBuilder.Definition.Namespace)
+
+		// Perform the update
+		result, err := testBuilder.Update()
```
Comment (trewest) [inline pkg/secret/secret_test.go:392]: It doesn't appear that you have actually updated the object with anything

File: pkg/secret/secret_test.go [hunk 19]
```diff
@@ -1 +1,516 @@
...
+}
+
+func buildSecretWithDummyObject() []runtime.Object {
+	return append([]runtime.Object{}, &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      defaultSecretName,
+			Namespace: defaultSecretNamespace,
+		},
+	})
+}
```
Comment (trewest) [inline pkg/secret/secret_test.go:456]: Please move these to the bottom of the file

File: pkg/secret/secret_test.go [hunk 20]
```diff
@@ -1 +1,508 @@
...
+	defaultSecretName      = "test-name"
+	defaultSecretNamespace = "test-namespace"
+	defaultSecretType      = "test-secretType"
+)
+
+func TestSecretPull(t *testing.T) {
+	testCases := []struct {
+		secretName          string
+		secretNamespace     string
+		expectedError       bool
```
Comment (kononovn) [inline pkg/secret/secret_test.go:25]: You can remove this bool argument and use
if expectedErrorText != "" condition in test case.

File: pkg/secret/secret_test.go [hunk 21]
```diff
@@ -1 +1,508 @@
...
+		if testCase.addToRuntimeObjects {
+			runtimeObjects = append(runtimeObjects, &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      testCase.secretName,
+					Namespace: testCase.secretNamespace,
+				},
+			})
+		}
+
+		testSettings = clients.GetTestClients(clients.TestClientParams{
```
Comment (kononovn) [inline pkg/secret/secret_test.go:86]: This is bug. Please check condition below and remove this line

File: pkg/secret/secret_test.go [hunk 22]
```diff
@@ -1 +1,508 @@
...
+			assert.Equal(t, testCase.name, testSecretBuilder.Definition.Name)
+			assert.Equal(t, testCase.namespace, testSecretBuilder.Definition.Namespace)
+		}
+	}
+}
+
+func TestSecretCreate(t *testing.T) {
+	testCases := []struct {
+		testSecret    *Builder
+		expectedError error
```
Comment (kononovn) [inline pkg/secret/secret_test.go:159]: Please add client bool parameter and add 1 more scenario when client is nil.

File: pkg/secret/secret_test.go [hunk 23]
```diff
@@ -1 +1,508 @@
...
+			expectedErr: "'data' cannot be empty",
+		},
+	}
+
+	for _, testCase := range testCases {
+		var runtimeObjects []runtime.Object
+		testBuilder, _ := buildTestBuilderWithFakeObjects(runtimeObjects, defaultSecretName, defaultSecretNamespace)
+
+		// Convert the test case data to map[string][]byte
+		data := make(map[string][]byte)
```
Comment (kononovn) [inline pkg/secret/secret_test.go:295]: Please move this map to test case parameters instead of key,value.

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix few comments.

**Review by kononovn — APPROVED** [review]
/lgtm

**Review by trewest — APPROVED** [review]
/lgtm
